File: threatflow-middleware/app/models/workflow.py
"""
Pydantic models for request/response validation
Enhanced with conditional logic support for Phase 4
"""

from pydantic import BaseModel, Field
from typing import List, Dict, Any, Optional
from enum import Enum

class NodeType(str, Enum):
    """Supported node types in workflow"""
    FILE = "file"
    ANALYZER = "analyzer"
    CONDITIONAL = "conditional"  # NEW: Phase 4
    RESULT = "result"

class ConditionType(str, Enum):
    """Types of conditions for conditional nodes"""
    VERDICT_MALICIOUS = "verdict_malicious"
    VERDICT_SUSPICIOUS = "verdict_suspicious"
    VERDICT_CLEAN = "verdict_clean"
    ANALYZER_SUCCESS = "analyzer_success"
    ANALYZER_FAILED = "analyzer_failed"
    CUSTOM_FIELD = "custom_field"

class ConditionalData(BaseModel):
    """Configuration data for conditional nodes"""
    condition_type: ConditionType
    source_analyzer: str  # Which analyzer's results to check
    field_path: Optional[str] = None  # JSON path for custom field checks
    expected_value: Optional[Any] = None  # Expected value for comparison

class WorkflowNode(BaseModel):
    """Represents a single node in the workflow canvas"""
    id: str = Field(..., description="Unique node identifier")
    type: NodeType = Field(..., description="Node type")
    data: Dict[str, Any] = Field(default_factory=dict, description="Node configuration")
    position: Optional[Dict[str, float]] = None
    conditional_data: Optional[ConditionalData] = None  # NEW: Phase 4

class WorkflowEdge(BaseModel):
    """Represents a connection between nodes"""
    id: str
    source: str = Field(..., description="Source node ID")
    target: str = Field(..., description="Target node ID")
    sourceHandle: Optional[str] = None
    targetHandle: Optional[str] = None

class WorkflowRequest(BaseModel):
    """Request model for workflow execution"""
    nodes: List[WorkflowNode]
    edges: List[WorkflowEdge]

class JobStatusResponse(BaseModel):
    """Response model for job status polling"""
    job_id: int
    status: str
    progress: Optional[int] = None
    analyzers_completed: int = 0
    analyzers_total: int = 0
    results: Optional[Dict[str, Any]] = None

class AnalyzerInfo(BaseModel):
    """Analyzer metadata"""
    name: str
    type: str
    description: Optional[str] = None
    supported_filetypes: Optional[List[str]] = None
    observable_supported: Optional[List[str]] = None

File: threatflow-frontend/src/types/workflow.ts
/**
 * TypeScript type definitions for ThreatFlow
 * Based on React Flow v11 and IntelOwl API
 */

import { Node, Edge } from 'reactflow';

// Re-export Edge for use in other modules
export type { Edge } from 'reactflow';

// ============= Node Types =============

export enum NodeType {
  FILE = 'file',
  ANALYZER = 'analyzer',
  RESULT = 'result',
  CONDITIONAL = 'conditional',
}

export interface FileNodeData {
  label: string;
  file: File | null;
  fileName: string;
  fileSize: number;
  onFileSelect?: (nodeId: string, file: File) => void;
}

export interface AnalyzerNodeData {
  label: string;
  analyzer: string;
  analyzerType: 'file' | 'observable';
  description: string;
  onAnalyzerSelect?: (nodeId: string, analyzer: string) => void;
}

export interface ResultNodeData {
  label: string;
  jobId: number | null;
  status: string;
  results: any | null;
  error: string | null;
}

export interface ConditionalNodeData {
  label: string;
  conditionType: 
    | 'verdict_malicious'           // Check if malware detected
    | 'verdict_suspicious'          // Check if suspicious
    | 'verdict_clean'               // Check if clean
    | 'analyzer_success'            // Check if analyzer succeeded
    | 'analyzer_failed'             // Check if analyzer failed
    | 'field_equals'                // Check if field equals value
    | 'field_contains'              // Check if field contains value
    | 'field_greater_than'          // Check if field > value
    | 'field_less_than'             // Check if field < value
    | 'yara_rule_match'             // Check if YARA rule matched
    | 'capability_detected';        // Check if Capa capability detected
  
  sourceAnalyzer: string;
  fieldPath?: string;               // e.g., "report.pe_info.signature.valid"
  expectedValue?: any;              // Value to compare against
  operator?: 'equals' | 'contains' | 'greater_than' | 'less_than' | 'regex';
  executionResult?: boolean | null; // null = not executed, true = condition met, false = condition not met
}

// Union type for all node data
export type CustomNodeData = FileNodeData | AnalyzerNodeData | ResultNodeData | ConditionalNodeData;

// Extended Node types with custom data
export type FileNode = Node<FileNodeData>;
export type AnalyzerNode = Node<AnalyzerNodeData>;
export type ResultNode = Node<ResultNodeData>;
export type ConditionalNode = Node<ConditionalNodeData>;
export type CustomNode = FileNode | AnalyzerNode | ResultNode | ConditionalNode;

// Use Node<any> for workflow state to handle union types properly
export type WorkflowNode = Node<any>;

// ============= API Types =============

export interface AnalyzerInfo {
  id: number;
  name: string;
  type: 'file' | 'observable';
  description: string;
  supported_filetypes: string[];
  not_supported_filetypes: string[];
  observable_supported: string[];
  available: boolean;
  unavailable_reason?: string;
}

export interface AnalyzersSummary {
  available_count: number;
  unavailable_count: number;
  total_count: number;
  containers_detected: {
    core: boolean;
    malware_tools: boolean;
    apk_analyzers: boolean;
    advanced_analyzers: boolean;
    observable_analyzers: boolean;
  };
}

export interface AnalyzersResponse {
  available: AnalyzerInfo[];
  unavailable: AnalyzerInfo[];
  summary: AnalyzersSummary;
}

export interface JobStatusResponse {
  job_id: number;
  status: string;
  progress?: number;
  analyzers_completed: number;
  analyzers_total: number;
  results: any | null;
  has_conditionals?: boolean;
  stagerouting?: StageRouting[];
}

export interface StageRouting {
  stage_id: number;
  target_nodes: string[];
  executed: boolean;
  analyzers?: string[];
}

export interface ExecuteWorkflowResponse {
  success: boolean;
  job_id?: number; // For backwards compatibility
  job_ids?: number[]; // For conditional workflows
  analyzers?: string[];
  total_stages?: number;
  executed_stages?: number[];
  skipped_stages?: number[];
  has_conditionals?: boolean;
  stagerouting?: StageRouting[]; // NEW: Routing metadata for conditional workflows
  message: string;
}

// ============= Workflow Types =============

export interface WorkflowState {
  nodes: CustomNode[];
  edges: Edge[];
  selectedNode: CustomNode | null;
  executionStatus: 'idle' | 'running' | 'completed' | 'error';
  jobId: number | null;
  uploadedFile: File | null;
}

export interface WorkflowActions {
  addNode: (node: CustomNode) => void;
  updateNode: (id: string, data: Partial<CustomNodeData>) => void;
  deleteNode: (id: string) => void;
  addEdge: (edge: Edge) => void;
  deleteEdge: (id: string) => void;
  setSelectedNode: (node: CustomNode | null) => void;
  setExecutionStatus: (status: WorkflowState['executionStatus']) => void;
  setJobId: (jobId: number | null) => void;
  setUploadedFile: (file: File | null) => void;
  reset: () => void;
}

export type WorkflowStore = WorkflowState & WorkflowActions;
